### PR TITLE
feat: add consul discovery module

### PR DESCRIPTION
**Pull request by Fabriceli**
### Description

As I mentioned previously in #8371 , my team submit our `consul` discovery module

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by spacewander** [comment]
Please make the CI pass, thanks!

**Comment by Fabriceli** [comment]
> Please make the CI pass, thanks!

ok, i fixed it

**Comment by Fabriceli** [comment]
> Learn more.

Could you start the other pipeline?

**Comment by spacewander** [comment]
> > Learn more.
> 
> Could you start the other pipeline?

Done

File: docs/en/latest/discovery/consul.md [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+### Configuration for Consul
+
+First of all, we need to add following configuration in `conf/config.yaml` :
+
+```yaml
+discovery:
+  consul:
+    servers:
+      - "http://127.0.0.1:8500"
+      - "http://127.0.0.1:8600"
```
Comment (tzssangglass) [inline docs/en/latest/discovery/consul.md:39]: 8500 and 8600 are different consul DNS servers that provide different query information?

File: apisix/discovery/consul/init.lua [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+local _M = {
+    version = 0.1,
+}
+
+
+local function discovery_consul_callback(data, event, source, pid)
+    all_services = data
+    log.notice("update local variable all_services, event is: ", event,
+            "source: ", source, "server pid:", pid,
+            ", all services: ", core.json.encode(all_services, true))
```
Comment (tzssangglass) [inline apisix/discovery/consul/init.lua:54]:
```suggestion
            ", all services: ", core.json.delay_encode(all_services, true))
```

**Comment by tzssangglass** [comment]
pls fix doc lint

File: docs/en/latest/discovery/consul.md (hunk 1 above)
Comment (Fabriceli) [reply docs/en/latest/discovery/consul.md:39]: they are different clusters, i add some comment in consul.md

File: apisix/discovery/consul/init.lua (hunk 2 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

**Comment by Fabriceli** [comment]
> pls fix doc lint

done

**Comment by Fabriceli** [comment]
@spacewander @tzssangglass I had finished, and I had fixed all the CI error

**Comment by Fabriceli** [comment]
@spacewander @tzssangglass I had finished, and I had fixed all the CI error

**Comment by Fabriceli** [comment]
@spacewander @tzssangglass I had finished, and I had fixed all the CI error

**Comment by Fabriceli** [comment]
@tzssangglass Could you give some tips for fix this pipeline [build (ubuntu-20.04, linux_openresty, t/gm)](https://github.com/apache/apisix/actions/runs/3571445230/jobs/6004502287#logs) error, I didnt modified anything about the  `grpc_server_example`, thanks

**Comment by tzssangglass** [comment]
> Could you give some tips for fix this pipeline [build (ubuntu-20.04, linux_openresty, t/gm)](https://github.com/apache/apisix/actions/runs/3571445230/jobs/6004502287#logs) error, I didnt modified anything about the `grpc_server_example`, thanks

rerun CI

**Comment by Fabriceli** [comment]
> > Could you give some tips for fix this pipeline [build (ubuntu-20.04, linux_openresty, t/gm)](https://github.com/apache/apisix/actions/runs/3571445230/jobs/6004502287#logs) error, I didnt modified anything about the `grpc_server_example`, thanks
> 
> rerun CI

I think the Github pipeline something wrong happened, the 12 pipelines have been running for more than 2 hours...

File: apisix/discovery/consul/init.lua [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+        if error_info then
+            log.error("connect consul: ", consul_server.consul_server_url,
+                    " by sub url: ", consul_server.consul_sub_url,
+                    ", got result: ", json_delay_encode(result, true),
+                    ", with error: ", error_info)
+
+            if not retry_delay then
+                retry_delay = 1
+            else
+                retry_delay = retry_delay * 4
```
Comment (spacewander) [inline apisix/discovery/consul/init.lua:256]: Can we wrap the get+retry into a function?

File: apisix/discovery/consul/init.lua [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+        skip_service_map = core.table.new(0, #consul_conf.skip_services)
+        for _, v in ipairs(consul_conf.skip_services) do
+            skip_service_map[v] = true
+        end
+    end
+
+    local consul_servers_list, err = format_consul_params(consul_conf)
+    if err then
+        error(err)
+        return
```
Comment (spacewander) [inline apisix/discovery/consul/init.lua:382]: Can we remove the unreachable code here?

File: conf/config-default.yaml [hunk 5]
```diff
@@ -309,6 +309,29 @@ nginx_config:                     # config for render the template to generate n
 #    dump:                         # if you need, when registered nodes updated can dump into file
 #       path: "logs/consul_kv.dump"
 #       expire: 2592000            # unit sec, here is 30 day
+#  consul:
+#    servers:
+#      - "http://127.0.0.1:8500"
+#      - "http://127.0.0.1:8600"
+#    skip_services:                    # if you need to skip special keys
```
Comment (spacewander) [inline conf/config-default.yaml:316]: Special services?

File: docs/en/latest/discovery/consul.md [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+
+First of all, we need to add following configuration in `conf/config.yaml` :
+
+```yaml
+discovery:
+  consul:
+    servers:
+      - "http://127.0.0.1:8500"
+      - "http://127.0.0.1:8600"   # `http://127.0.0.1:8500` and `http://127.0.0.1:8600` are different clusters
+    skip_services:                # if you need to skip special keys
```
Comment (spacewander) [inline docs/en/latest/discovery/consul.md:40]: Ditto

File: Makefile [hunk 7]
```diff
@@ -283,8 +283,9 @@ install: runtime
 
 	$(ENV_INSTALL) -d $(ENV_INST_LUADIR)/apisix/discovery
 	$(ENV_INSTALL) apisix/discovery/*.lua $(ENV_INST_LUADIR)/apisix/discovery/
-	$(ENV_INSTALL) -d $(ENV_INST_LUADIR)/apisix/discovery/{consul_kv,dns,eureka,nacos,kubernetes,tars}
+	$(ENV_INSTALL) -d $(ENV_INST_LUADIR)/apisix/discovery/{consul_kv,consul,dns,eureka,nacos,kubernetes,tars}
 	$(ENV_INSTALL) apisix/discovery/consul_kv/*.lua $(ENV_INST_LUADIR)/apisix/discovery/consul_kv
+	$(ENV_INSTALL) apisix/discovery/consul/*.lua $(ENV_INST_LUADIR)/apisix/discovery/consul
```
[the remark this answers is not in the document]
Comment (Fabriceli) [reply Makefile:287]: done

File: apisix/discovery/consul/init.lua [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+    end
+
+    log.info("read dump file: ", data)
+    data = util.trim(data)
+    if #data == 0 then
+        log.error("dump file is empty")
+        return
+    end
+
+    local entity, err  = core.json.decode(data)
```
[the remark this answers is not in the document]
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua (hunk 3 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua (hunk 4 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:382]: the `format_consul_params` will return error if the config uri is not valid

File: conf/config-default.yaml (hunk 5 above)
Comment (Fabriceli) [reply conf/config-default.yaml]: done

File: docs/en/latest/discovery/consul.md (hunk 6 above)
Comment (Fabriceli) [reply docs/en/latest/discovery/consul.md]: done

**Comment by Fabriceli** [comment]
@spacewander @tzssangglass Updated, please check again, thanks

File: apisix/discovery/consul/init.lua (hunk 4 above)
Comment (spacewander) [reply apisix/discovery/consul/init.lua:382]: @Fabriceli 
Yes. I mean if the `error` is hit, the `return` here is not executed.

Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

**Comment by Fabriceli** [comment]
@spacewander @tzssangglass the [CI](https://github.com/apache/apisix/actions/runs/3590288740/jobs/6047153281) with the t/xds-library/config_xds_2.t TEST 7 is not stable, I did not modified anything about that

**Comment by Fabriceli** [comment]
@spacewander @tzssangglass May you RETURN CI to rerun the CI again, thanks

**Comment by spacewander** [comment]
@Fabriceli 
Could you follow the discussion in https://github.com/apache/apisix/pull/8433#issuecomment-1334709384?
Thanks!

**Comment by Fabriceli** [comment]
> @Fabriceli Could you follow the discussion in [#8433 (comment)](https://github.com/apache/apisix/pull/8433#issuecomment-1334709384)? Thanks!

DONE, I had merge upstream master to dev branch.

**Comment by Fabriceli** [comment]
> @Fabriceli Could you follow the discussion in [#8433 (comment)](https://github.com/apache/apisix/pull/8433#issuecomment-1334709384)? Thanks!

**Comment by Fabriceli** [comment]
@spacewander @tzssangglass I have merged upstream master, could you re-start the CI? Thanks

File: apisix/discovery/consul/init.lua [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+
+    local resp_list = all_services[service_name]
+
+    if not resp_list then
+        log.error("fetch nodes failed by ", service_name, ", return default service")
+        return default_service and {default_service}
+    end
+
+    log.info("process id: ", ngx.worker.id(), ", all_services[", service_name, "] = ",
+            json_delay_encode(resp_list, true))
```
Comment (tzssangglass) [inline apisix/discovery/consul/init.lua:77]: it is better to use

```
local ngx_worker_id = ngx.worker.id

...

log.info("process id: ", ngx_worker_id(), ", all_services[", service_name, "] = ",
            json_delay_encode(resp_list, true))
```

File: apisix/discovery/consul/init.lua [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+end
+
+
+local function parse_instance(node)
+    local service_name, host, port = node.Service, node.Address, node.Port
+    -- if exist, skip special service name
+    if service_name and skip_service_map[service_name] then
+        return false
+    end
+    return true, host, tonumber(port), "", service_name
```
Comment (tzssangglass) [inline apisix/discovery/consul/init.lua:89]: what is the usage of `""` here?

File: t/discovery/consul.t [hunk 11]
```diff
@@ -0,0 +1,581 @@
...
+    "PUT /consul2/deregister/service_b2",
+    "PUT /consul1/register\n" . "{\"ID\":\"service_a1\",\"Name\":\"service_a\",\"Tags\":[\"primary\",\"v1\"],\"Address\":\"127.0.0.1\",\"Port\":30511,\"Meta\":{\"service_a_version\":\"4.0\"},\"EnableTagOverride\":false,\"Weights\":{\"Passing\":10,\"Warning\":1}}",
+    "PUT /consul1/register\n" . "{\"ID\":\"service_a2\",\"Name\":\"service_a\",\"Tags\":[\"primary\",\"v1\"],\"Address\":\"127.0.0.1\",\"Port\":30512,\"Meta\":{\"service_a_version\":\"4.0\"},\"EnableTagOverride\":false,\"Weights\":{\"Passing\":10,\"Warning\":1}}",
+    "PUT /consul1/register\n" . "{\"ID\":\"service_b1\",\"Name\":\"service_b\",\"Tags\":[\"primary\",\"v1\"],\"Address\":\"127.0.0.1\",\"Port\":30513,\"Meta\":{\"service_b_version\":\"4.1\"},\"EnableTagOverride\":false,\"Weights\":{\"Passing\":10,\"Warning\":1}}",
+    "PUT /consul1/register\n" . "{\"ID\":\"service_b2\",\"Name\":\"service_b\",\"Tags\":[\"primary\",\"v1\"],\"Address\":\"127.0.0.1\",\"Port\":30514,\"Meta\":{\"service_b_version\":\"4.1\"},\"EnableTagOverride\":false,\"Weights\":{\"Passing\":10,\"Warning\":1}}",
+]
+--- response_body eval
+["", "", "", "", "", "", "", "", "", "", "", ""]
```
Comment (tzssangglass) [inline t/discovery/consul.t:147]: check `--- error_code eval` is more clear?

File: t/discovery/consul.t [hunk 12]
```diff
@@ -0,0 +1,581 @@
...
+    "GET /hello",
+    "GET /hello",
+]
+--- response_body_like eval
+[
+    qr/server [1-2]\n/,
+    qr/server [1-2]\n/,
+]
+--- no_error_log
+[error, error]
```
Comment (tzssangglass) [inline t/discovery/consul.t:173]: what does this mean?  and cc @spacewander

Comment (spacewander) [reply t/discovery/consul.t:173]: A special `no_error_log` for pipelined_requests

File: apisix/discovery/consul/init.lua (hunk 10 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:89]: for metadata, which specifies arbitrary KV metadata linked to the service instance.
Added comment in the `init.lua` code

File: apisix/discovery/consul/init.lua (hunk 9 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: apisix/discovery/consul/init.lua (hunk 10 above)
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua:91]: done

File: apisix/discovery/consul/init.lua [hunk 13]
```diff
@@ -0,0 +1,416 @@
...
+
+
+local function update_all_services(server_name_prefix, data)
+    local sn
+    local up_services = core.table.new(0, #data)
+    local weight = default_weight
+    for _, node in pairs(data) do
+        local succ, ip, port, metadata, server_name = parse_instance(node)
+        if succ then
+            sn = server_name
```
[the remark this answers is not in the document]
Comment (Fabriceli) [reply apisix/discovery/consul/init.lua]: done

File: t/discovery/consul.t (hunk 11 above)
Comment (Fabriceli) [reply t/discovery/consul.t]: done

**Comment by tzssangglass** [comment]
> Some error in Chaos Test, May you give some hint about that

rerun it

**Review by tzssangglass — APPROVED** [review]

**Comment by Fabriceli** [comment]
Please take a look at this CR, thanks, cc @spacewander

**Pull request merged**